### PR TITLE
added a lint against function item references

### LINKED ISSUE #75239: Printing function address is confusing due to per-function ZST "singleton type"
I tried this code: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=61fe688172759b7756e84751557e39ea

```rust
use std::mem::size_of;
fn bar() -> usize { 42 }
type Left = [u8; 16];
println!("{:p}", size_of::<Left>);
println!("{:p}", bar);
```

*I expected to see this happen*:
The compilation succeed or guide me to
 - (1) cast the `fn` to `*const ()` with `as`.
 - (2) borrow the `fn`: `&size_of::<Left>`. But in this case the addresses printed are the same (what?).
    ```
    0x55867559f000
    0x55867559f000
    ```

*Instead, this happened*:
Cryptic error message:
```rust
error[E0277]: the trait bound `fn() -> usize {std::mem::size_of::<[u8; 16]>}: std::fmt::Pointer` is not satisfied
 --> src/main.rs:6:22
  |
6 |     println!("{:p}", size_of::<Left>);
  |                      ^^^^^^^^^^^^^^^ the trait `std::fmt::Pointer` is not implemented for `fn() -> usize {std::mem::size_of::<[u8; 16]>}`
  |
  = note: required by `std::fmt::Pointer::fmt`
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

There is `impl<usize> Pointer for fn() -> usize`: https://doc.rust-lang.org/nightly/std/fmt/trait.Pointer.html#impl-Pointer-3 .
It took me minutes before I found the workaround. But it still surprised me.

### Meta

`rustc --version --verbose`:
```
rustc 1.47.0-nightly (22ee68dc5 2020-08-05)
binary: rustc
commit-hash: 22ee68dc586440f96b76b32fbd6087507c6afdb9
commit-date: 2020-08-05
host: x86_64-unknown-linux-gnu
release: 1.47.0-nightly
LLVM version: 10.0
```

### PR DESCRIPTION
**Pull request by ayrtonm**
this lint suggests casting function references to `*const ()`
closes #75239 
r? @RalfJung

### CODE REVIEW

**Comment by jyn514** [comment]
Please add a test for this in `src/test/ui`. There are instructions at https://rustc-dev-guide.rust-lang.org/tests/adding.html#ui.

File: compiler/rustc_lint/src/builtin.rs
Comment (jyn514) [inline compiler/rustc_lint/src/builtin.rs:2437]: Hmm, why is this only looking in the current crate? I'd think we'd want to link on `&std::env::var` or something too.

**Comment by ollie27** [comment]
Shouldn't we suggest casting to a proper function pointer type rather than `*const ()`?

**Comment by lzutao** [comment]
> Shouldn't we suggest casting to a proper function pointer type

Sometimes function pointer type is really long.

**Comment by ollie27** [comment]
> Sometimes function pointer type is really long.

We could always suggest using underscores to prevent the types getting too long e.g. `as fn(_, _) -> _`.

**Comment by ayrtonm** [comment]
The lint breaks the following tests, so does it make sense to just add `#![allow(function_references)]` to not clutter them or should we add warn annotations?

```
custom-test-frameworks-simple.rs
enum-discriminant/actually_not_an_enum-discriminant.rs
functions-closures/fn-item-type-zero-sized.rs
issues/issue-16922-rpass.rs
issues/issue-20847.rs
issues/issue-22933-1.rs
issues/issue-25757.rs
trivial_casts-rpass.rs
unboxed-closures/unboxed-closures-extern-fn-hr.rs
unboxed-closures/unboxed-closures-extern-fn.rs
```

Also `compiler/rustc_typeck/src/check/mod.rs:776:37` triggers the lint, so I'm guessing there should be an allow attribute on that as well?

File: compiler/rustc_lint/src/builtin.rs
Comment (ayrtonm) [reply compiler/rustc_lint/src/builtin.rs:2437]: Good point, checking references to functions in other crates is probably important. I'm not sure I understand how to check other crates though. I've been looking through the docs for `hir::map::Map` and `ty::TyCtxt`, but if someone could give me a few pointers or help out that be great.

Comment (jyn514) [reply compiler/rustc_lint/src/builtin.rs:2437]: I don't think this needs to work on the HIR at all. Try [`type_of`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TyCtxt.html#method.type_of) instead: if this is both a function and a path, that means you're taking a reference to it. `type_of` returns a [`Ty`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/type.Ty.html), so you could call `matches!(ty.kind, TyKind::FnDef(_, _))`: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/enum.TyKind.html

File: src/test/ui/lint/function-references.rs
Comment (jyn514) [inline src/test/ui/lint/function-references.rs:12]:
```suggestion
    println!("{:p}", &baz);
    //~^ WARN cast `baz` with `as *const fn(_, _) -> _` to use it as a pointer
    println!("{:p}", &std::env::var);
    //~^ WARN cast `std::env::var` with `as *const fn(_, _) -> _` to use it as a pointer
```

**Comment by jyn514** [comment]
> Also compiler/rustc_typeck/src/check/mod.rs:776:37 triggers the lint, so I'm guessing there should be an allow attribute on that as well?

That code looks like this:

```rust
fn adt_destructor(tcx: TyCtxt<'_>, def_id: DefId) -> Option<ty::Destructor> {
    tcx.calculate_dtor(def_id, &mut dropck::check_drop_impl)
}
```
and [`calculate_dtor`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TyCtxt.html#method.calculate_dtor) takes `&mut dyn FnMut(Self, DefId) -> Result<(), ErrorReported>`. Maybe this shouldn't lint if it's being used as a `dyn Fn*`?

File: compiler/rustc_lint/src/builtin.rs
Comment (ollie27) [inline compiler/rustc_lint/src/builtin.rs:2443]:
```suggestion
                                        "cast `{}` with `as fn({}) -> _` to use it as a pointer",
```

We definitely shouldn't suggest a pointer to a function pointer 😉.

It might also be worth including things like ABI, unsafety and whether it's a variadic function so the suggestion could be applicable in more cases.

File: compiler/rustc_lint/src/builtin.rs
Comment (RalfJung) [inline compiler/rustc_lint/src/builtin.rs:2434]: I do not think it is necessary to restrict this to `hir::BorrowKind::Ref` -- all `AddrOf` should be handled the same. This makes sure that the lint also covers `ptr::raw_const!(function)`, which the test should make sure actually works.

File: compiler/rustc_lint/src/builtin.rs
Comment (RalfJung) [reply compiler/rustc_lint/src/builtin.rs:2437]: Yes, I think @jyn514 is right -- going through HIR here is unexpected.

File: src/test/ui/lint/function-references.stderr
Comment (RalfJung) [inline src/test/ui/lint/function-references.stderr:5]: I think the span should include the `&`, as it is part of the problem.

File: src/test/ui/lint/function-references.rs
Comment (RalfJung) [inline src/test/ui/lint/function-references.rs:15]: These are raw pointers to function pointers... that is certainly not what we want to test here. Please remove the `*const`.

File: compiler/rustc_lint/src/builtin.rs
Comment (RalfJung) [reply compiler/rustc_lint/src/builtin.rs:2437]: Actually, this is a design decision... do we want the lint to fire on
```rust
let fn_item = function_name;
let _val = &fn_item;
```
If yes, this should be type-based. If no, we have to instead check if the referent is the name of a function.

**Comment by RalfJung** [comment]
> The lint breaks the following tests, so does it make sense to just add `#![allow(function_references)]` to not clutter them or should we add warn annotations?
> 
> ```
> custom-test-frameworks-simple.rs
> enum-discriminant/actually_not_an_enum-discriminant.rs
> functions-closures/fn-item-type-zero-sized.rs
> issues/issue-16922-rpass.rs
> issues/issue-20847.rs
> issues/issue-22933-1.rs
> issues/issue-25757.rs
> trivial_casts-rpass.rs
> unboxed-closures/unboxed-closures-extern-fn-hr.rs
> unboxed-closures/unboxed-closures-extern-fn.rs
> ```
> 
> Also `compiler/rustc_typeck/src/check/mod.rs:776:37` triggers the lint, so I'm guessing there should be an allow attribute on that as well?

`actually_not_an_enum-discriminant.rs` should have the lint allowed for the line that triggers it, that looks deliberate.

16922, 20847, 22933, 25757 all seem deliberate... and they make me wonder if we need to improve the lint. They are all of the form `&fn_name as &dyn Something`, but sometimes the `as` is implicit. Taking a reference here is crucial. The same thing is probably happening with custom test frameworks, and it also explains the `rustc_typeck` failure.

So given that we have at least 5 false positives in the rustc codebase and test suite, we cannot leave the lint as-is, we have to find something smarter. The key thing that happens in all these cases is that the `&{fn item}` gets coerced or cast to an `&dyn Trait`; that should be something that we can detect, I think... the only problem is that to detect this we have to look "outside" of the `ExprKind::AddrOf`, and I do not know how to do that.

@oli-obk @Manishearth summoning some clippy people that can hopefully help with suggestions for how to implement such a lint. :) Basically we want to detect `&expr` where `expr` satisfies some conditions (we got that part covered) *and* the reference is not immediately cast/coerced to `&dyn Trait` -- how can we implement the latter check?

**Comment by oli-obk** [comment]
We could change the lint to a MIR lint that triggers on a `CastKind::Misc` where the casted value's type is a reference to a function item zst. This would ensure that we only ever lint the cast and never the creation.

**Comment by RalfJung** [comment]
We don't really have MIR lints though, do we? We have a few ad-hoc constructions that plug into different parts of the MIR pipeline, see https://github.com/rust-lang/rust/issues/72515.

**Comment by oli-obk** [comment]
The way clippy does this is to have a normal lint that loads the mir of any function definition it sees and then analyzes that. Having a good point in the optimization pipeline is something @rust-lang/wg-mir-opt should come up with. Until we have made an informed decision, just putting it next to https://github.com/rust-lang/rust/blob/f76eda3f01ba8e9714b579dc3e563ee973ade83d/compiler/rustc_mir/src/transform/mod.rs#L386 should be fine

**Comment by RalfJung** [comment]
Shouldn't we rather add it to one of the three places where we already have a MIR lint? E.g.

https://github.com/rust-lang/rust/blob/f76eda3f01ba8e9714b579dc3e563ee973ade83d/compiler/rustc_mir/src/transform/mod.rs#L309

**Comment by oli-obk** [comment]
Oh there they are :D I thought we had that and didn't find it and then thought I imagined it. Yea, that's a good place to get deterministic MIR

**Comment by RalfJung** [comment]
Not sure what you mean by "deterministic".^^

**Comment by oli-obk** [comment]
Optimizations can change things in unexpected ways, we don't want to optimize out some code on the current platform and then only lint on other platforms.

**Comment by ayrtonm** [comment]
It's still very much a WIP, but I changed lint to work with MIR. It now excludes `&fn_name as &dyn Trait` (implicit and explicit) and `(&fn_name)()` (function calls). This first MIR implementation emits a lint for
```
let fn_item = function_name;
let _val = &fn_item;
```
but I think it makes more sense to exclude it in this case. I also think we should exclude it when a function reference appears alone on the rhs of a regulat let binding like `let zst_ref = &function_name`. This broke one or two existing tests where `zst_ref` was just used by casting/coercion later on. It'd be nice to get a second opinion on this though.

I don't quite understand how to work with generics/`ParamEnv` in the compiler, so the coercion that happens in the first argument of `parameterized_call_fn` below doesn't exclude the lint yet.
```
fn parameterized_call_fn<F: Fn(u32) -> u32>(f: &F, x: u32) { f(x); }
parameterized_call_fn(&fn_name, 1);
```
It's also why almost all of these tests are failing.
```
consts/const_constructor/const-construct-call.rs#const_fn
consts/const_constructor/const-construct-call.rs#min_const_fn
enum-discriminant/actually_not_an_enum-discriminant.rs
functions-closures/fn-item-type-zero-sized.rs
issues/issue-20847.rs
lint/function-references.rs
mir/mir_calls_to_shims.rs
nll/promoted-bounds.rs
nll/user-annotations/adt-tuple-struct-calls.rs
nll/user-annotations/promoted-annotation.rs
struct-ctor-mangling.rs
unboxed-closures/unboxed-closures-extern-fn-hr.rs
unboxed-closures/unboxed-closures-extern-fn.rs
```

I also had another minor doubt. This implementation compares the spans of function references with function calls/casts/coercions to decide what function references to exclude. This works, but feels kinda janky... is there a better way to check if two MIR rvalues refer to the same thing in the source?

File: compiler/rustc_mir/src/transform/function_references.rs
Comment (ayrtonm) [inline compiler/rustc_mir/src/transform/function_references.rs:142]: Another minor question, is `skip_binder()` appropriate here and on line 144 in this context?

**Review by oli-obk — COMMENTED** [review]
>  I also think we should exclude it when a function reference appears alone on the rhs of a regular let binding like `let zst_ref = &function_name`. This broke one or two existing tests where `zst_ref` was just used by casting/coercion later on. It'd be nice to get a second opinion on this though.

Good point. So, since we're only trying to avoid casts from references to fn defintions to raw pointers or ints, we may be able to simplify the logic down to just looking at `Rvalue::Cast` and nothing else. Not sure if that works, let me know what you think:

* get the type of the operand in https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/enum.Rvalue.html#variant.Cast
* if that type is `TyKind::Ref` and that inner type is `TyKind::FnDef`, then we have a candidate
* if the `CastKind` from the `Rvalue::Cast` is `PointerCast::Unsize`, then we don't lint, in all other cases we lint

Is this missing some cases that we should lint or linting in cases that we should not lint?

**Comment by RalfJung** [comment]
> I also think we should exclude it when a function reference appears alone on the rhs of a regular let binding like let zst_ref = &function_name. 

I guess it is reasonable to start conservative and lint less; we can crank it up later if we feel we should.

> Is this missing some cases that we should lint or linting in cases that we should not lint?

Will this lint on `println!("{:p}", &fun);`, which explicitly came up in the original issue?

File: compiler/rustc_mir/src/transform/function_references.rs
Comment (RalfJung) [inline compiler/rustc_mir/src/transform/function_references.rs:147]: I would maybe say something like "to obtain a function pointer" or so.

File: compiler/rustc_session/src/lint/builtin.rs
Comment (RalfJung) [inline compiler/rustc_session/src/lint/builtin.rs:560]:
```suggestion
    "suggest casting to a function pointer when attempting to take references to function items",
```

File: compiler/rustc_session/src/lint/builtin.rs
Comment (RalfJung) [inline compiler/rustc_session/src/lint/builtin.rs:558]: The name might need some bikeshedding... maybe `FUNCTION_ITEM_REFERENCES` or so would be clearer? At least it is more precise.^^

**Comment by ayrtonm** [comment]
> > Is this missing some cases that we should lint or linting in cases that we should not lint?
> 
> Will this lint on `println!("{:p}", &fun);`, which explicitly came up in the original issue?

No, it would miss that as well as `format!` and `transmute` since no casting happens. I was thinking something along the lines of adding another exception to the lint for `StatementKind::Assign` when the lhs `Place` corresponds to a `LocalDecl` declared by the user and its type is a function reference.

File: compiler/rustc_mir/src/transform/function_references.rs
Comment (RalfJung) [inline compiler/rustc_mir/src/transform/function_references.rs:40]: I don't think I have seen a lint before that would do anything at the basic block level... why is this needed? There should be comments in the code explaining anything that is unusual, and this definitely is. :)

File: compiler/rustc_mir/src/transform/function_references.rs
Comment (RalfJung) [inline compiler/rustc_mir/src/transform/function_references.rs:51]: All these other functions have a `Location` which you can use with [this function](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/struct.Body.html#method.source_info) to obtain a `SourceInfo`. So I don't think you need to carry that around in the visitor.

**Comment by RalfJung** [comment]
> I also had another minor doubt. This implementation compares the spans of function references with function calls/casts/coercions to decide what function references to exclude. This works, but feels kinda janky... is there a better way to check if two MIR rvalues refer to the same thing in the source?

Yes, check the local they are based on -- or just directly compare them for equality.

> I don't quite understand how to work with generics/ParamEnv in the compiler, so the coercion that happens in the first argument of parameterized_call_fn below doesn't exclude the lint yet.

Ah, this is a tricky one... here `F` gets instantiated to the function type so we actually want the reference. :/
I honestly have no good idea for how to do a lint that doesn't warn in this case. We *do* have to take a reference to that function because the callee requires an `&F`. We'd have to check the trait bounds to realize that this is required, unlike in the printing case where there is no `Fn` trait involved.

So... I cannot think of anything better than what @oli-obk proposed, to check on the cast. But will that even catch any interesting cases?

**Comment by RalfJung** [comment]
@ayrtonm how would you like to proceed here?

Sorry this turned out to be much harder than I expected.^^

**Comment by ayrtonm** [comment]
Sorry for the delay, I got a bit caught up with work recently. I tried checking for an `Fn` trait when passing function references as arguments, but there are some valid cases where the argument isn't bound by any trait (like `std::mem::size_of_val` in an existing test). I found it's easier to just specifically check for `std::fmt::Pointer::<T>::fmt` and `std::mem::transmute::<T,_>` where the parameter type `T` is a function reference. It also checks all function calls for arguments bound by `Pointer` to warn about calls to functions like
```
fn print_ptr<F: Pointer>(f: F) { println!("{:p}", f); }
```
as well as cases that could potentially be a problem in the future like
```
fn bound_by_ptr_trait<F: Pointer>(f: F) { }
```

This doesn't break any existing tests, but it doesn't handle the following case yet
```
fn implicit_ptr_trait<F>(f: &F) { println!("{:p}", f); }
```
This compiles because `&F` always implements `Pointer` but I'm still thinking about how to lint in this case without affecting things like `size_of_val`.

In the working lints, I think the spans could be made clearer when passing arguments that aren't just a a single function reference like in `transmute::<_, (usize, usize)>((&foo, &bar))` where all of `(&foo, &bar)` is currently highlighted for both warnings. Maybe this other case could be made clearer as well.
```
warning: cast `foo` with `as fn() -> _` to use it as a pointer
  --> function-references.rs:67:22
   |
67 |     println!("{:p}", zst_ref);
   |                      ^^^^^^^
```
How would I go about making it also point out the span where `zst_ref` is defined within the same warning?


Also the function reference check is type-based, so 
```
let fn_item = function_name;
let _val = &fn_item;
```
triggers the lint. Even though `&fn_item` isn't the same as `&function_name`, I could see how someone might incorrectly think that the former is the address of `function_name`. I think it makes sense to lint, but let me know if anyone has any thoughts on this.

**Comment by RalfJung** [comment]
This all sounds great (including going type-based), and also I think instead of making the lint super perfect and catching all the possible cases, I'd first land a basic lint that is useful even if it misses some problems. That will also make review much easier. So maybe hold off on the `implicit_ptr_trait` thing, I don't think that is likely to occur in practice anyway.

**Comment by ayrtonm** [comment]
> I'd first land a basic lint that is useful even if it misses some problems.

Fair enough, I removed the test for `implicit_ptr_trait` and tweaked the lint wording like you suggested earlier.

**Comment by jyn514** [comment]
> Fair enough, I removed the test for implicit_ptr_trait and tweaked the lint wording like you suggested earlier.

Rather than removing it, can you mark it as `// ignore`? That way it's still around even if it's not implemented.

File: compiler/rustc_mir/src/transform/function_references.rs
Comment (oli-obk) [inline compiler/rustc_mir/src/transform/function_references.rs:58]: Can you pull this else block out into a separate function? The nesting is getting a bit much for my taste

**Comment by oli-obk** [comment]
Some general nits:

* Comments in the code base (mostly?) have a space after the `//` and some folks do try to start commments with a capitalized letter.
* Please leave an empty line between functions (and potentially throw a doc comment on the function if it isn't trivially clear what they do from their name)

**Comment by RalfJung** [comment]
r? @oli-obk 
(you seem to already be reviewing this anyway and you have clippy experience, so I hope this is okay. I am still happy to also take a look at this, but all this expression matching code is new to me.^^)

**Comment by oli-obk** [comment]
Sorry, I did not see that force push. Please leave a message when you force push

File: compiler/rustc_session/src/lint/builtin.rs
Comment (oli-obk) [reply compiler/rustc_session/src/lint/builtin.rs:2677]: This has not been addressed yet

File: compiler/rustc_mir/src/transform/function_item_references.rs
Comment (oli-obk) [inline compiler/rustc_mir/src/transform/function_item_references.rs:186]: The lint message should contain what is wrong, not what should be done to fix it. Since the fix is likely automatically applicable, you can use [`span_suggestion`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/diagnostic_builder/struct.DiagnosticBuilder.html#method.span_suggestion) before `emit` in order to produce an actual suggestion that rust-analyzer or `cargo fix` can apply (semi-)automatically

**Review by oli-obk — CHANGES_REQUESTED** [review]
Just a few small things, then this is ready to go.

**Comment by ayrtonm** [comment]
> Sorry, I did not see that force push. Please leave a message when you force push

My bad, I didn't realize you wouldn't get a notification.

So the lint message could  be something like "taking a reference to a function item does not give a function pointer" and the suggestion could be "cast `ident` to obtain a function pointer: `ident as fn()`"?

**Comment by oli-obk** [comment]
> So the lint message could be something like "taking a reference to a function item does not give a function pointer" and the suggestion could be "cast `ident` to obtain a function pointer: `ident as fn()`"?

yes. that looks right.

**Comment by ayrtonm** [comment]
Just made the changes @oli-obk requested. I left the suggestion's applicability `Unspecified` instead of `MaybeIncorrect` since in cases like `transmute((&foo,))`, the exact substitution won't result in valid Rust code. The problem is that getting the subset of the span that should be replaced from MIR isn't straightforward, but these cases should be rare in practice.

**Review by oli-obk — APPROVED** [review]

**Comment by oli-obk** [comment]
@bors r+

Thanks!

**Pull request merged**
